org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/DetectVMInstallationsJob.java

```diff
@@ -0,0 +1,121 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jdt.launching.IVMInstall;
+import org.eclipse.jdt.launching.IVMInstallType;
+import org.eclipse.jdt.launching.JavaRuntime;
+import org.eclipse.jdt.launching.VMStandin;
+import org.eclipse.osgi.util.NLS;
+
+class DetectVMInstallationsJob extends Job {
+
+	private static final Object FAMILY = "Family " + DetectVMInstallationsJob.class.getName(); //$NON-NLS-1$
+	
+	DetectVMInstallationsJob() {
+		super(LaunchingMessages.lookupInstalledJVMs);
+	}
+
+	@Override
+	protected IStatus run(IProgressMonitor monitor) {
+		Collection<File> rootDirectories = new HashSet<>();
+		if (!Platform.OS_WIN32.equals(Platform.getOS())) {
+			rootDirectories.add(new File("/usr/lib/jvm")); //$NON-NLS-1$
+		}
+		rootDirectories.add(new File(System.getProperty("user.home"), ".sdkman/candidates/java")); //$NON-NLS-1$ //$NON-NLS-2$
+		String javaHome = System.getenv("JAVA_HOME"); //$NON-NLS-1$
+		if (javaHome != null) {
+			rootDirectories.add(new File(javaHome));
+		}
+		String jdkHome = System.getenv("JDK_HOME"); //$NON-NLS-1$
+		if (jdkHome != null) {
+			rootDirectories.add(new File(jdkHome));
+		}
+		// other common/standard lookup strategies can be added here
+
+		Collection<File> candidates = rootDirectories.stream().filter(File::isDirectory)
+			.map(dir -> dir.listFiles(File::isDirectory))
+			.filter(Objects::nonNull)
+			.flatMap(Arrays::stream)
+			.map(t -> {
+				try {
+					return t.getCanonicalFile();
+				} catch (IOException e) {
+					return null;
+				}
+			}).filter(Objects::nonNull)
+			.collect(Collectors.toCollection(HashSet::new));
+		if (monitor.isCanceled()) {
+			return Status.CANCEL_STATUS;
+		}
+		Set<File> knownInstallations = Stream.of(JavaRuntime.getVMInstallTypes())
+			.map(IVMInstallType::getVMInstalls)
+			.flatMap(Arrays::stream)
+			.map(IVMInstall::getInstallLocation)
+			.filter(Objects::nonNull)
+			.map(t -> {
+				try {
+					return t.getCanonicalFile();
+				} catch (IOException e) {
+					return null;
+				}
+			}).filter(Objects::nonNull)
+			.collect(Collectors.toSet());
+		candidates.removeIf(knownInstallations::contains);
+		StandardVMType standardType = (StandardVMType)JavaRuntime.getVMInstallType(StandardVMType.ID_STANDARD_VM_TYPE);
+		candidates.forEach(f -> {
+			if (monitor.isCanceled()) {
+				return;
+			}
+			new Job(NLS.bind(LaunchingMessages.configuringJVM, f.getAbsolutePath())) {
+				@Override
+				public IStatus run(IProgressMonitor monitor) {
+					VMStandin workingCopy = new VMStandin(standardType, f.getAbsolutePath());
```

Comment:
Somewhere along this code a check is missing that skips non-valid "JRE" directories which aren't actually JRE's. Just export  /tmp in JDK_HOME and enjoy various crazy "JRE" directories found.

After next startup they disappear but surely we won't confuse anyone in case some garbage input is accepted here as "JRE".

I believe `org.eclipse.jdt.internal.launching.StandardVMType.validateInstallLocation(File)` is needed.